tests/test_vec_envs.py

```diff
@@ -62,6 +62,14 @@ def custom_method(dim_0=1, dim_1=1):
         return np.ones((dim_0, dim_1))
 
 
+def test_vecenv_func_checker():
+    """The functions in ``env_fns'' must return distinct instances since we need distinct environments."""
+    env = CustomGymEnv(gym.spaces.Box(low=np.zeros(2), high=np.ones(2)))
+
+    with pytest.raises(ValueError):
```

Comment:
although it should be the same, I would also add a test with `make_vec_env(lambda: env, n_envs=2)` as it is the other common mistake I think.

Reply:
You have a nose for it.

The next test does not pass.

```python
env = gym.make("CartPole-v1")

with pytest.raises(ValueError):
    make_vec_env(lambda: env, n_envs=2)

env.close()
```

 I am investigating.

Reply:
i guess because of the monitor wrapper?
you have to use unwrapped to get the original env

Reply:
i guess because of the monitor wrapper?
you have to use unwrapped to get the original env

Reply:
Ok... It's because it's wrapped, so the id is the different, even though the inner env is the same. @Rocamonde, can you fix this?

```python
import gym
from stable_baselines3.common.env_util import make_vec_env

env = gym.make("CartPole-v1")

print(id(env))  # 4551646560

venv = make_vec_env(lambda: env, n_envs=2)

print([id(env) for env in venv.envs])  # [5311397600, 5311397648]
print([id(env.env) for env in venv.envs])  # [4551646560, 4551646560]

env.close()
```

Reply:
`env.unwrapped` is what you are looking for, but you need to check that this method exist first.

Reply:
Hmm. What do we actually want to do here, raise an error if any two levels of nesting match? I think we only care about the object at the top of the hierarchy. If the instances at the top are distinct, then that's a guarantee that all wrappers are distinct, since you can't a same wrapper instance wrapping two different environment instances.

In think we should write a helper that extracts the primitive environment for each environment, then compare those ids. Does that sound sensible?

Reply:
>  I think we only care about the object at the top of the hierarchy.

we care about the original gym env, which is returned by `unwrapped`.

> If the instances at the top are distinct, then that's a guarantee that all wrappers are distinct, since you can't have a same wrapper instance wrapping two different environment instances.

not sure to understand.
But for short, if the gym instance is the same, even though the wrappers are different, there is an issue (see example in the tests).

> In think we should write a helper that extracts the primitive environment for each environment,

this is what you get with the `unwrapped` attribute ;)

Reply:
Wdym by "almost"? Why is it an attribute that sometimes does not exist? Surely you should have a function that maps `f(any_env: Union[gym.Env, vec_env.VecEnv]) -> unwrapped_env: gym.Env`

Reply:
Yes, agreed that this is an issue. I was trying to say that you need to unwrap the env all the way, since that's what contains the stateful attributes. And that two envs having different id's when fully unwrapped always implies that any wrappers have different id's, but not the other way around (since the user could still be using the same gym env but then e.g. creating different wrapper instances with different params).

Reply:
> Why is it an attribute that sometimes does not exist? 

it always exists (it is in the base class for gym env and wrappers).

> EDIT: this helper almost exist actually

I meant we have helpers to unwrap `VecEnv` and `gym.Env`:  https://github.com/DLR-RM/stable-baselines3/blob/7e1db1aaaa0f486cceb79faf5a08a25c5fded670/stable_baselines3/common/env_util.py#L11

> still be using the same gym env but then e.g. creating different wrapper instances with different params).

How would that not create issues?
Because at the end, you are still updating the same env instance despite different wrappers.

Reply:
> Great will use that in the PR!

there might be a misunderstanding, `id(env.unwrapped)` is what you should use, no need for helper